### DISCUSSION TITLE
Creating a LuaSnip snippets plugin

### DISCUSSION THREAD

**Discussion by evesdropper** [L3MON4D3/LuaSnip#822] [General]
Hi! Not sure if this is the right place to ask, but I got some advice to publish my snippets into a plugin. However, I'm not sure how to create a plugin that loads the snippets i have and utils files. Some questions:

1. First of all, does anyone else have an interest in having LaTeX snippets widely available in one click (basically install and go like friendly-snippets but with better snippets)? It seems like it would be useful.
2. If so, is there anyone with plugin development experience who would be willing to help me develop this plugin?

Thanks!

**Comment by L3MON4D3** [comment]
Nice idea!
IMO it's high time there is a plugin with high-quality snippets (in general, but latex is the filetype which imo benefits the most from thought-out snippets)

Maybe basic plugin-stuff:
* I think you'd only need to have the `<yourplugindirectory>/lua/<plugin-name>/init.lua`, in which you load the snippets. The `init.lua` can then be accessed/executed via `require("<plugin-name>")`

I've had some thoughts about the issues of writing a plugin with lua-snippets for some time, and I'm going to just put them here, I hope they're useful to you :sweat_smile:

* The most feature-rich way of loading snippets is via the loaders (they can: edit files contributing snippets, reload on edit (not that important in this case I guess), lazy-load snippets (which is very nice to have)), BUT it might be a bit difficult to only load the snippets contributed by a plugin, since `load` needs either an absolute path (not sure if there's a 100% reliable way of getting the location of the plugin) or just loads any `luasnippets`-directory it can find in runtimepath, but this would also mean loading snippets that don't belong to the plugin (potentially at least).
One solution for this would be for us to add a setting where the name of the directory that is searched in runtimepath can be adjusted, then the snippets could be stored in some very specific directory like `<yourplugindirectory>/evesdropper_latex_snippets_dont_touch`, and load only that (sorry for getting all technical, I hope my point came across :sweat_smile:).
* when using the loaders, `snip_env` (which is responsible for getting all the single-letter-shorthands into the snippet files) can't be relied on, it might be different on other machines. So, it can either be not used at all, or we have to add another option to `load`, where the `snip_env` can be set for the files loaded by this call (this feels like the nicer option, I'll have to check how feasible it is to implement this)

If your plugin won't contain a great amount of snippets, you could also just use `add_snippets`, this would be much more straightforward, but lazy-load and the other niceties are lost.

I don't think I want to really closely work on this plugin, but if you have specific questions, feel free to open a discussion here (or in your plugins github ofc, and ping me) :)

**Reply by evesdropper** [reply]
Thanks for the advice! I've never worked on plugin development before, so getting started especially right now is a bit overwhelming. Hopefully this all comes to fruition sometime in the summer.

> The most feature-rich way of loading snippets is via the loaders

So all I would technically need to do is:

```lua
require("luasnip.loaders.from_lua").load({ paths = "<>" })
```
where `<>` is the directory in which snippets are stored (abstracting away the technicalities of loading).

Perhaps it could support both the plugin snippets and personal snippets using a config variable that changes the value of `paths`? Maybe something to think about in the future; I still have no idea how to develop a plugin :sweat_smile: 

The `snip-env` bit I also noticed when checking out [this plugin](https://github.com/molleweide/LuaSnip-snippets.nvim); it seems like it would be needed to be loaded on every file, which would be pretty inefficient. Having something that just loads the shorthands would be really nice if possible; otherwise, I think maybe I could just put `snip-env` in a utils file and require it with every snippet file.

I'm hoping that I can expand the snippet collection a bit more after LaTeX, so using a loader over `add_snippets` is probably the best solution.

**Reply by leiserfg** [reply]
One can always do 
```lua
local here = debug.getinfo(1, "S").source:sub(2)
```
at the top of a lua file to get the path. 
Putting that in the  `plugin/choose_a_better_name.lua` , modifying the output accordingly and calling `require("luasnip.loaders.from_lua").load({ paths = the_final_path })`  with it should do it.

**Reply by leiserfg** [reply]
We could even expose a function for doing that (if more snippets collections start appearing) :thinking:

**Reply by no-vici** [reply]
> The `snip-env` bit I also noticed when checking out [this plugin](https://github.com/molleweide/LuaSnip-snippets.nvim); it seems like it would be needed to be loaded on every file, which would be pretty inefficient. Having something that just loads the shorthands would be really nice if possible; otherwise, I think maybe I could just put `snip-env` in a utils file and require it with every snippet file.

I only load the snip-env once with all the shortcuts enclosed in it, and in other [filetype].lua I just return {...}. I even threw your get_visual() in there and call it in any [filetype].lua and it works perfectly.

**Reply by evesdropper** [reply]
Yeah, this is true. However, I think L3MON4D3 is trying to say that it may be different when considering something you install vs. for a personal collection of snippets.

**Reply by no-vici** [reply]
> Yeah, this is true. However, I think L3MON4D3 is trying to say that it may be different when considering something you install vs. for a personal collection of snippets.

I think the asset that LuaSnip provides is to let users write their own snippets, each programmer has his/her own taste and habbits. This is why LuaSnip didn't come with pre-written snippets and users come up with their own creativities. I have my own collection of snippets written in LuaSnip's formats, and the truth is others might not like mine. Turning your collection into a plugin to me is not worth the effort, you could share yours and help people learn how to make their own ones. I'm an old man and came to computer quite late in life, still be able to figure out my ways through LuaSnip's docs and examples, I'm sure that young people can do it better than I did, and I have to admit that I've learned a lot from reading your blog. So big thanks.

**Reply by evesdropper** [reply]
That is true, and I agree that it would be best if you pick up a new technology and be familiar with it, especially since you are able to create something that aligns (near) perfectly with your preferences. But to play devil's advocate, not everyone has the desire/ability/need to spend many hours poring through documentation or guides nor has the time/energy to come up with snippets (and to that extent, configurations in general), especially when they need to construct a solution rather quickly. And for those reasons, some preconfiguration may be helpful.

In fact, this is why a good fraction of the snippets I have now are pretty much a transcription/slight modification of [Gilles Castel's LaTeX snippets in UltiSnips](https://github.com/gillescastel/latex-snippets/) and that a lot of the plugins I have configured are either out of the box or configurations designed mostly by other users and barely (if at all) modified for my own usage. I needed something quick so that I would be able to use Neovim in class to take notes in LaTeX. I learned a lot about LuaSnip due to the lack of documentation/out of the box setup (friendly-snippets just did not mesh with me at all for the reasons you described), and honestly I really enjoyed it and do not regret in the slightest. However, there may be plenty of other people like me, who might want some solution that they can implement over the course of a couple hours, not over a couple weeks, or maybe they are familiar with another plugin like UltiSnips and want to have a more seamless transition to LuaSnip without having to give up their current snippets.

I hope to create a collection that allows users to have some common snippets (e.g. from UltiSnips, and perhaps a few of my own), and then from there maybe supplement my snippets with their own - saving users time but also allowing them to customize the experience to their own. Hence why I was thinking to add something which allows users to have their own snippets alongside common snippets provided by the plugin, and hopefully in the future also creating a feature that allows users to switch triggers for provided snippets.

**Reply by no-vici** [reply]
> [...] And for those reasons, some preconfiguration may be helpful.

I can't agree more with this from a newbie's perspective. However, instead of veering off with another plugin, would it be more centralised to merge with LuaSnip as a goody? Or providing users with your settings, what they need to do is plugging in the lua file ( require"evesdropper") and snippet-collection, then voila.

> I hope to create a collection that allows users to have some common snippets (e.g. from UltiSnips, and perhaps a few of my own), and then from there maybe supplement my snippets with their own - saving users time but also allowing them to customize the experience to their own. Hence why I was thinking to add something which allows users to have their own snippets alongside common snippets provided by the plugin, and hopefully in the future also creating a feature that allows users to switch triggers for provided snippets.

How about [snippet-converter](https://github.com/smjonas/snippet-converter.nvim)? The user can convert their snippets in UltiSnips into Luasnip with a few keystrokes right?

**Reply by L3MON4D3** [reply]
> Hopefully this all comes to fruition sometime in the summer

:crossed_fingers: :crossed_fingers:

> So all I would technically need to do is:

Yup, exactly :+1:

> Perhaps it could support both the plugin snippets and personal snippets using a config variable that changes the value of paths?

Not sure what should be configurable about the value of paths :sweat_smile:
Adding "load from custom-named rtp-directory" could also be done with a separate parameter, like `rtp_dirname`

> I think maybe I could just put snip-env in a utils file and require it with every snippet file

Yep :+1: That would definitely be more efficient than declaring all the locals in every file, downside ofc. that  you'd need to do `locals.s, locals.t` (or more likely `l.s, l.t`) (not a big downside though)

> I'm hoping that I can expand the snippet collection a bit more after LaTeX

I was kind of hoping to hear that xD

> One can always do ... at the top of a lua file to get the path.

Ouh, didn't know we could get debug-stuff structured like that, pretty cool :+1:
(Tangent: we could use this to get the line where a snippet was defined and somehow integrate this with metadata (#719), such that we could can not only jump to the file, but also to the line where some snippet is defined)

> We could even expose a function for doing that (if more snippets collections start appearing) 

Maybe something like `path_relative` as an argument to `load`? Since we are able to get the location of the caller from `debug.getinfo` this should be straightforward to implement

> I think L3MON4D3 is trying to say that it may be different when considering something you install vs. for a personal collection of snippets

:ok_hand:

> Hence why I was thinking to add something which allows users to have their own snippets alongside common snippets provided by the plugin

Selectively disabling/overriding the provided snippets provided would be an awesome feature :+1::+1:
Switching triggers sounds very ambitious, but it's definitely possible

**Reply by evesdropper** [reply]
> The most feature-rich way of loading snippets is via the loaders (they can: edit files contributing snippets, reload on edit (not that important in this case I guess), lazy-load snippets (which is very nice to have)), BUT it might be a bit difficult to only load the snippets contributed by a plugin, since load needs either an absolute path (not sure if there's a 100% reliable way of getting the location of the plugin) or just loads any luasnippets-directory it can find in runtimepath, but this would also mean loading snippets that don't belong to the plugin (potentially at least).
One solution for this would be for us to add a setting where the name of the directory that is searched in runtimepath can be adjusted, then the snippets could be stored in some very specific directory like <yourplugindirectory>/evesdropper_latex_snippets_dont_touch, and load only that (sorry for getting all technical, I hope my point came across sweat_smile).

Hmm yeah, I just started working on this and I ran into this problem. I'm not sure what the second paragraph really means, could you elaborate a bit more? @L3MON4D3

**Reply by L3MON4D3** [reply]
That's the one about loaders&paths, right?
The issue is that loaders needs an absolute path, but the path depends on where the plugin is eventually installed.
Luckily, `debug.getinfo(1, "Sl")` should print the location of the file calling it, so going from there it should be possible to find the directory which contains the snippets

**Reply by evesdropper** [reply]
Hmm, that makes sense. it's still quite difficult to get something done due to the lack of ability to test things :sob:

I think under that logic `require("luasnip.loaders.from_lua").load(debug.getinfo(1).source:sub(2) .. "luasnippets")` should give the full file path, but nothing is loading either. Is there any way to see more clearly how things are being detected?

**Reply by L3MON4D3** [reply]
Check the log, there should be some info in there (call `ls.log.set_loglevel("info")` before calling load)
But the issue might just be that `load` wants a table with the `paths`-key set to the paths, so `load({paths = debug.getinfo(1).source:sub(2) .. "luasnippets"})` might do it (also, is there a `/` missing? And removal of `.lua`?)

**Reply by evesdropper** [reply]
Hmm, fixing the table wasn't it unfortunately. I think there shoudn't be a `/` missing or need to remove `.lua` to my knowledge, though I might be wrong (at least that's what I found online when checking how `debug.getinfo` worked).

Logs say
```
ERROR | lua-loader: Failed to execute
      | : /home/revise/dotfiles/nvim/lua/snippets/tex/tex.lua
ERROR | lua-loader: Failed to execute
      | : /home/revise/dotfiles/nvim/lua/snippets/tex/tex.lua
```
which seems a bit orthogonal :sob:

**Reply by L3MON4D3** [reply]
Mhmmm, yeah that seems to be a different issue😅
I'm pretty sure debug.getinfo returns the filename it was executed in. I'm not sure how exactly your file-layout looks like, but I just assumed that the snippets aren't located in something like `init.lualuasnippets`

**Reply by evesdropper** [reply]
Hm, file structure looks like this:
```
 .
├──  LICENSE
├──  lua
│   └──  luasnip-latex-snippets
│       ├──  init.lua
│       └──  luasnippets
│           └──  tex
│               ├──  commands.lua
│               ├──  delimiters.lua
│               ├──  environments.lua
│               ├──  math-symbols.lua
│               ├──  math.lua
│               ├──  preamble.lua
│               └──  utils
│                   ├──  conditions.lua
│                   └──  init.lua
└──  README.md
```
and from [this link](https://neovim.discourse.group/t/get-path-to-plugin-directory/2658) I think this should give something of the form `lua/luasnip-latex-snippets/`, so just appending `luasnippets` would seem to make sense. (Also tried adding `:gsub("init.lua", "")` in case. Not quite sure though, I'm honestly a terrible programmer so I might've just made a huge leap of logic :sweat_smile: 

Finally got some logs though:
```
INFO  | lua-loader: Loading lazy-load-snippets for filetype `tex`
INFO  | lua-loader: Loading lazy-load-snippets for filetype `cpp`
INFO  | lua-loader: Loading lazy-load-snippets for filetype `python`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `tex`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `cpp`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `python`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `tex`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `cpp`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `python`
INFO  | lua-loader: Loading lazy-load-snippets for filetype ``
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype ``
INFO  | vscode-loader: Loading lazy-load-snippets for filetype ``
INFO  | lua-loader: Loading lazy-load-snippets for filetype ``
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype ``
INFO  | vscode-loader: Loading lazy-load-snippets for filetype ``
INFO  | lua-loader: Loading lazy-load-snippets for filetype `lua`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `lua`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `lua`
```

**Reply by L3MON4D3** [reply]
Ah, okay full path is a bit ambiguous, might also include the filename. Best way to be sure is to `vim.pretty_print` it I'd say
Bit surprised, the logs should also include messages like "loading snippets at ...", maybe I'm mistaken..

**Reply by L3MON4D3** [reply]
Ah nvm, we don't print those if the directory does not exist :sweat_smile:

**Reply by evesdropper** [reply]
Hmm, wasn't sure how to `vim.print`, but at least using lua's built in print I got `init.lua`, so I'd assume the `gsub` bit would account for it. 

Just loading the snippets seems to be the hardest part :pensive:

**Reply by evesdropper** [reply]
> Ah nvm, we don't print those if the directory does not exist sweat_smile

I assume you mean the log (and that it doesn't pick up on the snippet folder)?

**Reply by L3MON4D3** [reply]
Yeah, if the directory is correct we will add an `info`-message to the log, otherwise nothing

> Hmm, wasn't sure how to vim.print, but at least using lua's built in print I got init.lua, so I'd assume the gsub bit would account for it.

Right, `vim.pretty_print` is deprecated
`vim.print(<what-you-would-otherwise-pass-to-load>)`, it will pretty-print tables and the like..
Best print the result of `:gsub`, that should make the issue more obvious

> Just loading the snippets seems to be the hardest part :disappointed:

Weeelll, should be smooth sailing from there though (:crossed_fingers:)

**Reply by evesdropper** [reply]
Hmm, can't figure out how to `vim.print` :sob:

But with regular print I get

```
lua/luasnip-latex-snippets/	1
```
which almost seems like what I want, except there's an extra 1.

**Reply by L3MON4D3** [reply]
Ahhhh, the 1 should be the second return-value from gsub.
It should work fine if you only store the first return-value, like
```lua
local path = debug.getinfo(1, "Sl").source:gsub("init.lua", "luasnippets")
require("luasnip.loaders.from_lua").lazy_load({paths = path})
```
(This behaves the same as `local path, _ = ...:gsub(...)`)

**Reply by evesdropper** [reply]
Hmm,  yeah that makes sense. Rip though, I'm unable to figure out why it doesn't work as the logs don't show a filename still :cry:

**Reply by L3MON4D3** [reply]
Is the path correct? Could you check by printing it?

**Reply by evesdropper** [reply]
Yeah, I checked earlier (mb), shows `lua/luasnip-latex-snippets/luasnippets` as intended I believe.

**Reply by L3MON4D3** [reply]
Oh, I think the `init.lua` is just not executed xD
You can add something like `require("luasnip-latex-snippets")` in `plugin/luasnip-latex-snippets.lua` (*.lua in `<rtp>/plugin/` will be sourced on startup, see `:h startup`)

**Reply by evesdropper** [reply]
oh my god, that's so unfortunate :sob: thank you for figuring out this rookie mistake

this sort of debugging gives me throwbacks to taking data structures

and well yes, it does load now and it works, thank you so much for tolerating my silly mistakes.

~~[company which hired me], you did not see any of this~~

**Reply by L3MON4D3** [reply]
That's the kind of stuff you either know or you don't :sweat_smile:
Now you do :D

**Comment by kiryph** [comment]
> First of all, does anyone else have an interest in having LaTeX snippets widely available in one click (basically install and go like friendly-snippets but with better snippets)? It seems like it would be useful.

Yes, I think this is useful. It can be quite time-consuming to create a satisfying set of LaTeX snippets.

Do you know following?

Plugins

- Iuri Mateus https://github.com/iurimateus/luasnip-latex-snippets.nvim (first commit: Dec 18, 2021)
- JC Zhang https://github.com/jczhang02/luasnips-mathtex-snippets (first commit: Sep 5, 2022)

Personal configs with comprehensive LaTeX snippets

- Elijan Mastnak https://github.com/ejmastnak/dotfiles/tree/main/config/nvim/LuaSnip/tex with a tutorial
  https://www.ejmastnak.com/tutorials/vim-latex/luasnip/
- Hashem A. Damrah (SingularisArt) https://github.com/SingularisArt/snippets/tree/master/tex (UltiSnips)

**Reply by evesdropper** [reply]
Hmm yeah, I've seen them all. I've also got a [repo of my own](https://github.com/evesdropper/dotfiles/tree/main/nvim/lua/snippets/tex) as well as a set of [guides](https://evesdropper.dev/files/luasnip/), hence it would be nice to ship off some documentation and added defaults to Castel's snippets.

So pretty much, I already have most of the snippets I plan to use, though setting things up the way I want it will probably be not as easy :sob:

**Reply by kiryph** [reply]
What I miss most about these projects is kind of a cheatsheet similar to https://michaelgoerz.net/refcards/vimlatexqrc.pdf and https://ftp.gnu.org/pub/gnu/auctex/11.88-extra/tex-ref.pdf

Even when using a well-crafted set of snippets, it would be great to have a summary which lists them.

**Reply by evesdropper** [reply]
:eyes: I could definitely add those - shouldn't take too much time to generate if I can get a script that lists triggers, descriptions and docstrings.

**Reply by L3MON4D3** [reply]
:point_right::point_right: `require("luasnip.extras.snippet_list").open()` :point_left::point_left:

**Reply by evesdropper** [reply]
Thanks :sweat_smile: thought about that but forgot I could pass custom functions in. Is there some way where I could get the output of `snippet_list` to maybe save to a file?

**Reply by L3MON4D3** [reply]
Mhmm I think it's possible to extend it to do just that (look at the callbacks it supports), but just copy-paste from that buffer sounds easier.
Or do you want to do some automation with it?

**Reply by evesdropper** [reply]
Yeah, automation -  thinking that every time I push an update or something it can rewrite.

**Reply by L3MON4D3** [reply]
Ah, okay.. Yeah you can do that by modifying `display`, its input is that string which should be put into in a buffer, but you could instead save it to a file.
(You might even modify `printer` such that valid latex is produced??)

**Reply by evesdropper** [reply]
I had the following first:
```lua
-- Print Snippet Documentation
local sl = require("luasnip.extras.snippet_list")

local display_snippet_info = function (printer_result)
    info_path = debug.getinfo(1).source:sub(2):gsub("lua/luasnip-latex-snippets/init.lua", "doc/snippets.txt")
    local f = assert(io.open(info_path, "w"))
    f:write(printer_result)
    f:close()
end

sl.open({display = display_snippet_info})
```
though I'm not sure how this would display or get formatted into documentation.

I was thinking of using a bash script to get this done instead. Are there any other more "neovim" like ideas?

**Reply by L3MON4D3** [reply]
Yeah, getting that into documentation might be more complicated...
(Although, modifying printer such that a markdown-list-entry is produced for each snippet should be possible, only you'd then also have to add some preamble in `display_snippet_info`)

One issue with `snippet_list` is that it explicitly looks at all available snippets, not only those loaded by your plugin.
A more robust approach might be to explicitly restrict this to the snippets loaded by your plugin, although that is not yet possible (the association sourcefile->snippet is stored by us, but it's not yet accessible via api)
(wouldn't worry about not going through `snippet_list` for this, its functionality can be recreated pretty quickly, and it will probably be more readable)

**Reply by evesdropper** [reply]
Ah gotcha, I'm probably thinking to just write my own for now just because it's not too much so far (and it gives me the liberty to adjust the verbosity of a description depending on the output - e.g. compress multiple related snippets into one explanation, go more into detail with another more involved one).

Will also probably see what I can do with `snippet_list` though later down the line - as always, thanks for the advice and information!

**Reply by L3MON4D3** [reply]
Yeah, handcrafted all the way, as long as the effort is worth it :+1:
Also, if the description is good, one can also just use `snippet_list` to get a quick overview of available snippets, but that will obviously not be structured properly, only by filetype

> as always, thanks for the advice and information!

Yeah, you're welcome :)
